### PR TITLE
Dump refined queries

### PR DESCRIPTION
**Pull request by karmacoma-eth**
Problem: previously, we would only dump unrefined queries.

By moving the dumping inside solve, we dump every assertion query, including refined queries.

### CODE REVIEW

File: src/halmos/__main__.py [hunk 1]
```diff
@@ -1015,13 +1007,22 @@ class GenModelArgs:
...
 
-def solve(query: str, args: Namespace) -> Tuple[CheckSatResult, Model]:
+def solve(
+    query: str, args: Namespace, dump_filename: Optional[str] = None
+) -> Tuple[CheckSatResult, Model]:
+    if dump_filename:
+        with open(dump_filename, "w") as f:
+            f.write("(set-logic QF_AUFBV)\n")
+            f.write(query)
+            f.write("(get-model)\n")
```
Comment (daejunpark) [inline src/halmos/__main__.py:1024]: `(get-model)` will be an error if the result is unsat.  indeed, z3 provides a command line option that automatically prints a model if sat.  if other smt solvers have such an option, we wouldn't need this.

File: src/halmos/__main__.py [hunk 2]
```diff
@@ -335,8 +335,9 @@ def run_bytecode(hexcode: str, args: Namespace) -> List[Exec]:
         else:
             print(f"Final opcode: {mnemonic(opcode)})")
             print(f"Return data: {returndata}")
+            dump_dirname = f"/tmp/halmos-{uuid.uuid4()}"
```
Comment (daejunpark) [inline src/halmos/__main__.py:338]: `uuid.uuid4().hex` to be consistent?

Comment (daejunpark) [inline src/halmos/__main__.py:338]:
```suggestion
            dump_dirname = f"/tmp/halmos-{uuid.uuid4().hex}"
```

File: src/halmos/__main__.py (hunk 1 above)
Comment (daejunpark) [reply src/halmos/__main__.py:1024]: turned out that some smt solvers don't provide such an option, so there is no clear solution that works for all solvers.  for now, we can keep this knowing that the query will generate error when unsat, and come back to this later if needed.

**Review by daejunpark — APPROVED** [review]

File: src/halmos/__main__.py (hunk 1 above)
Comment (karmacoma-eth) [reply src/halmos/__main__.py:1024]: sorry forgot to update this: I checked and all the solvers are tried are ok with `(get-model)` even in the unsat case. 

z3 does prints an error, but that's after correctly returning `unsat` so I think it's good overall. 

One thing I still need to check if `(get-model)` makes a big difference in the `sat` case

**Pull request merged**